### PR TITLE
Inserter: Display block by category

### PR DESCRIPTION
**Pull request by youknowriad**
- Adding the category to the block API
 - Adding `wp.blocks.getCategories`
 - For now, I'm not allowing "registering" new categories but a block should choose between the already defined categories, easily changeable if needed.
 - I'm adding `lodash` as a dependency, let me know if you have objections about this.

### CODE REVIEW

File: blocks/README.md [hunk 1]
```diff
@@ -125,6 +127,7 @@ Registers a new block provided a unique slug and an object defining its behavior
 
 - `title: string` - A human-readable [localized](https://codex.wordpress.org/I18n_for_WordPress_Developers#Handling_JavaScript_files) label for the block. Shown in the block picker.
 - `icon: string | WPElement` - Slug of the [Dashicon](https://developer.wordpress.org/resource/dashicons/#awards) to be shown in the control's button, or an element if you choose to render your own SVG.
+- `category: string` - Id of the block's category. The category is used to organize the blocks in the block inserter.
```
Comment (nylen) [inline blocks/README.md]: What is an `Id` in this context?  What format should it have?  Do we need to enforce/test this?

File: blocks/categories.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+
+/**
+ * Block categories
+ *
+ * @var {Array} categories
+ */
+const categories = [
+	{ id: 'common', label: 'Common' }
```
Comment (nylen) [inline blocks/categories.js]: Instead of `id` and `label`, I'm thinking we should use the same convention as blocks itself, which have a `slug` and a `title`.

**Comment by aduth** [comment]
Do we anticipate that a plugin author would want to register their own custom category? cc @jasmussen

File: editor/inserter/index.js [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
 		<div className="inserter">
 			<div className="inserter__arrow" />
 			<div className="inserter__content">
-				<div className="inserter__Category-blocks">
-					{ blocks.map( ( { slug, title, icon } ) => (
-						<div key={ slug } className="inserter__block">
-							<span className={ 'dashicons dashicons-' + icon } />
-							{ title }
+				{ categories
+					.filter( ( category ) => !! blocksByCategory[ category.id ] )
```
Comment (aduth) [inline editor/inserter/index.js]: Ever a time there's multiple iterators like here, I'm inclined to nudge toward `Array#reduce` instead.

File: package.json [hunk 4]
```diff
@@ -50,6 +50,7 @@
     "webpack-node-externals": "^1.5.4"
   },
   "dependencies": {
-    "hpq": "^1.1.1"
+    "hpq": "^1.1.1",
+    "lodash": "^4.17.4"
```
Comment (aduth) [inline package.json]: Minor: If we bring this in, we'll want to think about (a) how/if we want to enforce Lodash vs. native usage (`forEach`, `assign`) and (b) if we think we'll ever need to reconcile with core's use of Underscore (maybe this always has a different build process?).

**Comment by jasmussen** [comment]
> Do we anticipate that a plugin author would want to register their own custom category? cc @jasmussen

If block plugins take off like we want them to, I'd expect them to really _want to_, but I don't know that we want to allow it. Think categories like "HyperCompuMegaCorp(R) SEO Pluginz(tm)" as a category. 

Hey, we may want to allow that. I don't know. But right at this moment it feels like we could provide a good list of stock categories that most plugins should be able to register them into. Such as:

- common
- gallery
- commerce
- charts/graphs
- social
- embeds
- other

That list may be too big, or not exhaustive enough, we'd probably want to tune it a bit as we go.

File: blocks/README.md (hunk 1 above)
Comment (youknowriad) [inline blocks/README.md]: I think we should validate the `settings` object and check the mandatory attributes (title, icon, category), but since it's not done yet for any setting. I'd leave this to its own PR.

File: editor/inserter/index.js (hunk 3 above)
Comment (youknowriad) [inline editor/inserter/index.js]: I could have swear you're going to make this comment :)

More seriously, I think we have two different coding styles :). You care a lot about Performance which is good, and maybe I care a bit more about making the code meaningfull (Functional approach).

In my code we could replace the code by 

```js
categories.filter(categoryHasBlocks).map(renderCategory)
```

and with a reduce I found it less meaningfull

```js
categories.reduce(concatRenderIfHasBlocks)
```

I would use the `reduce` if we really have performance issues, but I'll choose the first approach if we don't. I hopes this explains well how I approach this kind of code

File: package.json (hunk 4 above)
Comment (youknowriad) [inline package.json]: Could we add some technical guidelines to the docs of this project to explain some choices? React, JSX, how we use lodash, ... 

I'm ok not using `lodash` for now. ES6 features are not as "complete" as lodash, but we can do a lot without any lib.

Comment (aduth) [inline package.json]: >Could we add some technical guidelines to the docs of this project to explain some choices? React, JSX, how we use lodash, ...

We have a few notes about the first two in [Element docs](https://github.com/WordPress/gutenberg/tree/master/element). I agree it would be a good idea, even for our own benefit to ensure our choices are well reasoned.

File: editor/inserter/index.js [hunk 5]
```diff
@@ -1,18 +1,31 @@
...
 	return (
 		<div className="inserter">
 			<div className="inserter__arrow" />
 			<div className="inserter__content">
-				<div className="inserter__Category-blocks">
-					{ blocks.map( ( { slug, title, icon } ) => (
-						<div key={ slug } className="inserter__block">
-							<span className={ 'dashicons dashicons-' + icon } />
-							{ title }
+				{ categories
```
Comment (aduth) [inline editor/inserter/index.js]: Can we use `Array#reduce` for performance? (`O(n)` instead of `O(2n)` not counting inner `map`)

Comment (youknowriad) [inline editor/inserter/index.js]: Already answered this above :)

I think we have two different coding styles :). You care a lot about Performance which is good, and maybe I care a bit more about making the code meaningful (Functional approach).

In my code we could replace the code by

```js
categories.filter(categoryHasBlocks).map(renderCategory)
```

and with a reduce I found it less meaningful

```
categories.reduce(concatRenderIfHasBlocks)
```

I would use the reduce if we really have performance issues, but I'll choose the first approach if we don't. I hope this explains well how I approach this kind of code

Comment (aduth) [inline editor/inserter/index.js]: >Already answered this above :)

Whoops! I had sensed I'd already remarked but missed the collapsed comment when looking through the conversation history.

To your point: Fair to call it a difference of opinion, but aside from performance wins, I'd also disagree that `reduce` is any less legible. To me it's simply looping to aggregate a value with conditions on (or formatting applied to) the values included.

```js
categories.reduce( ( memo, category ) => {
	if ( !! blocksByCategory[ category.slug ] ) {
		memo.push(
			<div key={ category.slug } className="inserter__Category-blocks">
				<div className="inserter__separator">{ category.title }</div>
				{ blocksByCategory[ category.slug ].map( ( { slug, title, icon } ) => (
					<div key={ slug } className="inserter__block">
						<span className={ 'dashicons dashicons-' + icon } />
						{ title }
					</div>
				) ) }
			</div>
		);
	}

	return memo;
}, [] )
```

Comment (BE-Webdesign) [inline editor/inserter/index.js]: You would probably just want to do `reduce` and `pipe` if you broke each mapping filtering function into something that could be called.  That way you loop once and apply the same chain sequentially.  If that makes sense.

Comment (aduth) [inline editor/inserter/index.js]: I updated my previous example with a more compelling argument for `Array#reduce` preserving functionally composing the filter / map:

```js
const filterHasBlocks = () => /* ... */;
const mapRenderCategory = () => /* ... */;

categories.reduce( ( memo, category ) => (
	memo.concat( flow( filterHasBlocks, mapRenderCategory )( [ category ] ) )
), [] )
```

_(h/t @mrbobbybryant)_

**Comment by aduth** [comment]
If we pull in Lodash, we should also use [`babel-plugin-lodash`](https://www.npmjs.com/package/babel-plugin-lodash) to support cherry-picking used methods. Perhaps even [`lodash-webpack-plugin`](https://www.npmjs.com/package/lodash-webpack-plugin) in anticipation of some off-by-default features of the [roadmapped Lodash 5](https://github.com/lodash/lodash/wiki/Roadmap).

File: editor/inserter/index.js (hunk 5 above)
Comment (youknowriad) [inline editor/inserter/index.js]: I updated the code, since React don't render anything when we return "false", I'm just mapping over the categories

**Comment by youknowriad** [comment]
@aduth I was hoping that we won't need the plugin with webpack2, but apparently since `lodash` don't support the `module` entry on the `package.json` yet

**Comment by youknowriad** [comment]
@aduth I've used the lodash-babel-plugin but the webpack plugin broke my `groupBy` call, I left it out for now

File: package.json (hunk 4 above)
Comment (mtias) [inline package.json]: Let's leave out lodash for now. Agreed on starting to document decisions.

**Comment by youknowriad** [comment]
Dropped lodash for now, we'll see if we needed a lot.

File: editor/inserter/index.js [hunk 6]
```diff
@@ -1,18 +1,31 @@
...
+					.map( ( category ) => !! blocksByCategory[ category.slug ] && (
+						<div key={ category.slug } className="inserter__Category-blocks">
```
Comment (aduth) [inline editor/inserter/index.js]: I assume the capitalization on "Category" `className` was unintentional. Maybe also we ought to avoid adding classes until we start styling them?

**Comment by aduth** [comment]
>Dropped lodash for now, we'll see if we need it a lot.

My only worry is that we'll introduce all these verbose hand-written alternatives and risk not refactoring them when we inevitably choose to incorporate Underscore or Lodash.

**Comment by BE-Webdesign** [comment]
I am thumbs up on bringing in Lodash or Underscores sooner than later.  Is it possible to use Underscores version that ships with WordPress core?

**Review by aduth — APPROVED** [review]
We should plan to document available categories, but it's not clear whether this is settled at the moment. For now this LGTM 👍

**Comment by mtias** [comment]
Yes, the categories are definitely fuzzy at the moment, and this will potentially be reused by the customizer blocks, which will add more categories I presume.
